From 156ce8d8d9ee1b614b0079e1e9506af1835bdab7 Mon Sep 17 00:00:00 2001
From: Jacob Stiffler <j-stiffler@ti.com>
Date: Wed, 7 Nov 2018 15:48:47 -0500
Subject: [PATCH] execute_command: workaround for GUI apps with weston

* For all GUI apps, shutdown matrix before launching
  - This is needed since matrix is launched fullscreen, and another
    GUI app may get hidden behind matrix with no way to recover.
* For MM and 3D and null window apps, shutdown weston before launching
  - These apps use KMS sink or DRM, which collides with weston
  - Wait 1 second to make sure the application freeing all SGX resources
    prior to weston shutdown
Signed-off-by: Jacob Stiffler <j-stiffler@ti.com>
Signed-off-by: Eric Ruei <e-ruei1@ti.com>
---
 execute_command.sh | 40 +++++++++++++++++++++++++++++++++++++++-
 generate.php       |  4 ++++
 run_script.php     | 10 +++++++++-
 3 files changed, 52 insertions(+), 2 deletions(-)

diff --git a/execute_command.sh b/execute_command.sh
index 325dcb9..b8bb6da 100755
--- a/execute_command.sh
+++ b/execute_command.sh
@@ -63,7 +63,22 @@ do
 	outputfilename=$item 
   fi
 
-  if [ $i -gt 1 ]
+  if [ $i -eq 2 ]
+  then
+    program_type=$item
+  fi
+
+  if [ $i -eq 3 ]
+  then
+    category=$item
+  fi
+
+  if [ $i -eq 4 ]
+  then
+    win_type=$item
+  fi
+
+  if [ $i -gt 4 ]
   then
     touch "lock/"$item
 
@@ -72,10 +87,33 @@ do
 
 done
 
+if [ $program_type == gui ]
+then
+  /etc/init.d/matrix-gui-2.0 stop
+
+  if [ $category == 3d -o $category == multimedia -o $category == touch -o $win_type == null ]
+  then
+    sleep 1
+    /etc/init.d/weston stop
+    sleep 1
+  fi
+fi
+
 echo "Filename:"$filename
 echo "Output:"$outputfilename
 eval $filename > "tmp/"$outputfilename 2>&1
 
+if [ $program_type == gui ]
+then
+  if [ $category == 3d -o $category == multimedia -o $category == touch -o $win_type == null ]
+  then
+    /etc/init.d/weston start
+    sleep 1
+  fi
+
+  /etc/init.d/matrix-gui-2.0 start
+fi
+
 #Using a more unique string to detect if the script is completed
 echo "_?!!MATRIX_SCRIPT_COMPLETED!!?_" >> "tmp/"$outputfilename
 
diff --git a/generate.php b/generate.php
index 10cfed4..cd88dd3 100644
--- a/generate.php
+++ b/generate.php
@@ -105,6 +105,10 @@ for($x = 0;$x<count($contents)&&strlen($contents[$x])>0;$x++)
 		$top["Description_Link"] = get_contents("X-MATRIX-Description",$dotdesktop);
 
 		$top["Lock"] = get_contents("X-MATRIX-LOCK",$dotdesktop);
+
+		$top["Category"] = get_contents("Categories",$dotdesktop);
+
+		$top["WinType"] = get_contents("WinType",$dotdesktop);
 	} 
 
 	if($category == -1)
diff --git a/run_script.php b/run_script.php
index 9dd7a10..d34a286 100644
--- a/run_script.php
+++ b/run_script.php
@@ -55,6 +55,7 @@ $submenu = isset($_GET["submenu"]) == true ? $_GET["submenu"] : "main_menu" ;
 $found_app = get_application($var,$submenu,$_GET["app"]);
 
 $lock_list = $found_app["Lock"];
+$win_type = $found_app["WinType"];
 
 //Verify that there is a lock specified for this application
 if($lock_list != -1 )
@@ -91,7 +92,14 @@ if($currently_locked==false)
 	$random_string = strval(rand());
 	$random_string .= strval(rand());
 
-	$script_command = "./execute_command.sh \"".addslashes($script_link). "\" ".$random_string.".txt ".$lock_list;
+        if($win_type != -1 )
+	{
+		$script_command = "./execute_command.sh \"".addslashes($script_link). "\" ".$random_string.".txt \"".$found_app["ProgramType"]."\" \"".$found_app["Category"]."\" \"".$found_app["WinType"]."\"  ".$lock_list;
+	}
+	else
+	{
+		$script_command = "./execute_command.sh \"".addslashes($script_link). "\" ".$random_string.".txt \"".$found_app["ProgramType"]."\" \"".$found_app["Category"]."\"  \"wayland\" ".$lock_list;
+	}
 
 	$last_line = system($script_command." > /dev/null 2>/dev/null & ", $retval);
 }
-- 
1.9.1

